sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -248,7 +261,12 @@ def _dict_union(dicts):
   return result
 
 
-def _flatten(valueish, root=()):
+def _flatten(
+    valueish,  # type: Union[T, Tuple[T, ...], Dict[Any, T]]
```

Comment:
do we want to cover the `List[T]` case for `valueish`?  

Alternately, do we wan to make this more generic:

```python
def _flatten(
    valueish,  # type: Union[T, Iterable[T], Mapping[Any, T]]
    root=(),  # type: Tuple[Any, ...]
    ):
  # type: (...) -> Dict[Tuple[Any, ...], T]
  """Given a nested structure of dicts, tuples, and lists, return a flat
  dictionary where the values are the leafs and the keys are the "paths" to
  these leaves.

  For example `{a: x, b: (y, z)}` becomes `{(a,): x, (b, 0): y, (b, 1): c}`.
  """
  if isinstance(valueish, typing.Mapping):
    return _dict_union(_flatten(v, root + (k, )) for k, v in valueish.items())
  elif isinstance(valueish, typing.Iterable):
    return _dict_union(
        _flatten(v, root + (ix, )) for ix, v in enumerate(valueish))
  else:
    return {root: valueish}
```

Another thought, is it valid / worthwhile to create a relationship between `valueish`, `root`, and the result key?:

```python
def _flatten(
    valueish,  # type: Union[T, Iterable[T], Mapping[U, T]]
    root=(),  # type: Tuple[U, ...]
    ):
  # type: (...) -> Dict[Tuple[U, ...], T]
```

Reply:
I should mention that typing provides a lot of opportunity to go down rabbit-holes, so the right answer is often "it would be more accurate, but it's not that valuable".   A common motivator behind investing in accurately typing utility functions is when it allows you to avoid adding manual / explicit types or casts elsewhere in the code.   Imagine a scenario where mypy knows key of the `valueish` map, but after it passes through `_flatten`, you need to re-type the result because it becomes `Any`.

Reply:
Generally I'd agree that `Iterable`/`Mapping` would be preferable, but here I want to be restrictive about the kinds of values I decompose. 

The keys types of the mapping would be `Union[None, int, U]`, so you'd have to cast anyway, so I think it's simpler to leave as is. 

Still, good food for thought.